### PR TITLE
SBND PR for DQM v1 04 00 Release

### PR DESCRIPTION
**Pull request by wjdanswjddl**
Description

- updates to SBND DQM analysis modules for all subsystems: TPC, PMT, CRT, Timing, Trigger.
- updated fcls in subdirectories to test the individual modules, and `online_analysis.fcl` in the `installations/sbn-nd` that runs all subsystem streams 
- `sbndqm/dqmanalysis` TPC and CRT subdirectories are now separated in to ICARUS and SBND subdirectories to prevent conflicts with future developments. SBND modules were renamed by appending "SBND" in the module names

Testing details

All updates are specific to SBND, and were tested on online SBND runs.

### CODE REVIEW

File: sbndqm/dqmAnalysis/TPC/SBND/OnlineEvdSBND_module.cc [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "art_root_io/TFileService.h"
+
+#include "sbndqm/dqmAnalysis/TPC/ChannelDataSBND.hh"
+#include "sbndqm/Decode/TPC/HeaderData.hh"
+#include "sbndqm/dqmAnalysis/TPC/AnalysisSBND.hh"
+
+//#include "sbndqm/Decode/TPC/SBND/ChannelMap/TPCChannelMapService.h"
+#include "sbndcode/ChannelMaps/TPC/TPCChannelMapService.h"
```
Comment (mvicenzi) [inline sbndqm/dqmAnalysis/TPC/SBND/OnlineEvdSBND_module.cc]: I'm getting a compilation error from this include statement... we don't have `sbndcode` available at ICARUS.
I suppose the lines got mixed up? You want to include the local channel mapping service, right?

File: sbndqm/dqmAnalysis/TPC/SBND/OfflineEvdSBND_module.cc [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "art_root_io/TFileService.h"
+
+#include "sbndqm/dqmAnalysis/TPC/ChannelDataSBND.hh"
+#include "sbndqm/Decode/TPC/HeaderData.hh"
+#include "sbndqm/dqmAnalysis/TPC/AnalysisSBND.hh"
+
+// #include "sbndqm/Decode/TPC/SBND/ChannelMap/TPCChannelMapService.h"
+#include "sbndcode/ChannelMaps/TPC/TPCChannelMapService.h"
```
Comment (mvicenzi) [inline sbndqm/dqmAnalysis/TPC/SBND/OfflineEvdSBND_module.cc]: Same compilation error from here.

File: sbndqm/dqmAnalysis/TPC/SBND/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+cet_build_plugin( OnlineEvdSBND art::module
+        ROOT::Core
+        lardataobj::RecoBase
+        larcore::Geometry_Geometry_service
+        artdaq_utilities::artdaq-utilities_Plugins
+        sbndaq_online_storage
+        sbndaq_online_redis_connection
+        sbndaq_online_hiredis
+        sbndcode_ChannelMaps_TPC_TPCChannelMapService_service
```
Comment (mvicenzi) [inline sbndqm/dqmAnalysis/TPC/SBND/CMakeLists.txt]: We can't link to `sbndcode_ChannelMaps_TPC_TPCChannelMapService_service` at ICARUS.
Did you mean to use the local channel mapping service?

File: sbndqm/dqmAnalysis/TPC/SBND/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+cet_build_plugin( OfflineEvdSBND art::module
+        ROOT::Core
+        lardataobj::RecoBase
+        larcore::Geometry_Geometry_service
+        artdaq_utilities::artdaq-utilities_Plugins
+        sbndaq_online_storage
+        sbndaq_online_redis_connection
+        sbndaq_online_hiredis
+        sbndcode_ChannelMaps_TPC_TPCChannelMapService_service
```
Comment (mvicenzi) [inline sbndqm/dqmAnalysis/TPC/SBND/CMakeLists.txt]: Same here. We can't link to `sbndcode_ChannelMaps_TPC_TPCChannelMapService_service` at ICARUS

File: sbndqm/Decode/TPC/SBND/CMakeLists.txt [hunk 5]
```diff
@@ -1,19 +1,27 @@
...
-  ${ART_FRAMEWORK_CORE}
-  ${ROOT_BASIC_LIB_LIST}
+  art::Utilities
+  fhiclcpp::fhiclcpp
+  messagefacility::MF_MessageLogger
+  art::Framework_Core
+  ROOT::Core
+  ROOT::Tree 
+  #sbndqm_Decode_TPC_SBND_ChannelMap_TPCChannelMapService_service
+  sbndcode_ChannelMaps_TPC_TPCChannelMapService_service
```
Comment (mvicenzi) [inline sbndqm/Decode/TPC/SBND/CMakeLists.txt]: `sbndcode_ChannelMaps_TPC_TPCChannelMapService_service` can't be linked at ICARUS

File: sbndqm/Decode/TPC/SBND/ChannelMap/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+                           art::Framework_Core
+                           art::Persistency_Provenance
+                           messagefacility::MF_MessageLogger
+                           ROOT::Core
+         MODULE_LIBRARIES
+                           art_root_io::TFileService_service
+                           lardataobj::RawData
+                           larcore::Geometry_Geometry_service
+                           #sbndqm_Decode_TPC_SBND_ChannelMap_TPCChannelMapService_service 
+                           sbndcode_ChannelMaps_TPC_TPCChannelMapService_service
```
Comment (mvicenzi) [inline sbndqm/Decode/TPC/SBND/ChannelMap/CMakeLists.txt]: Same here. Can't be linked at ICARUS.

**Review by mvicenzi — CHANGES_REQUESTED** [review]
Unfortunately the current code is not compiling properly at ICARUS because we don't have `sbndcode` available. I'm getting an error regarding the following `include` statement coming from both the SBND event display modules:
```
/home/nfs/icarus/DQM_DevAreas/DQM_2024-09-22_MV_testv1_03_00/srcs/sbndqm/sbndqm/dqmAnalysis/TPC/SBND/OfflineEvdSBND_module.cc:38:10: fatal error: sbndcode/ChannelMaps/TPC/TPCChannelMapService.h: No such file or directory
   38 | #include "sbndcode/ChannelMaps/TPC/TPCChannelMapService.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
Even if I comment it out and I leave the `include` to the local channel map service (see below), I get the following link errors:
 ```
 /usr/bin/ld: cannot find -lsbndcode_ChannelMaps_TPC_TPCChannelMapService_service
collect2: error: ld returned 1 exit status
gmake[2]: *** [sbndqm/slf7.x86_64.e26.prof/lib/libsbndqm_dqmAnalysis_TPC_SBND_OfflineEvdSBND_module.so] Error 1
gmake[1]: *** [sbndqm/sbndqm/dqmAnalysis/TPC/SBND/CMakeFiles/sbndqm_dqmAnalysis_TPC_SBND_OfflineEvdSBND_module.dir/all] Error 2
```

If the plan was to switch to using the local channel map service in `sbndqm/Decode/TPC/SBND/ChannelMap`, it seems it's not working properly. The `include` statements and the `CMakeList` files should be changed to point to the local library instead of the one from `sbndcode`.

**Comment by wjdanswjddl** [comment]
@mvicenzi Updated so that the channel map is called from the local sbndqm subdirectory. Renamed scripts to not clash with sbndcode, as we still need sbndcode set up locally for geometry services.

**Review by mvicenzi — APPROVED** [review]
Thank you @wjdanswjddl !
I checked out the new changes and now it builds without issues.
I've also tested it running alongside `sbndaq v1_10_02` and I see no problems.
From the ICARUS side we can proceed to merge this in and tag the release.

I approve!

**Comment by wesketchum** [comment]
I got an error building:

```
[ 64%] Building CXX object sbndqm/sbndqm/dqmAnalysis/CRT/ICARUS/CMakeFiles/sbndqm_dqmAnalysis_CRT_ICARUS_BernCRTdqm_module.dir/BernCRTdqm_module.cc.o
In file included from /home/sbn/sbndqm_v1_03_00/srcs/sbndqm/sbndqm/dqmAnalysis/TPC/AnalysisSBND.hh:28,
                 from /home/sbn/sbndqm_v1_03_00/srcs/sbndqm/sbndqm/dqmAnalysis/TPC/SBND/OfflineEvdSBND_module.cc:35:
/home/sbn/sbndqm_v1_03_00/srcs/sbndqm/sbndqm/dqmAnalysis/TPC/FFT.hh:5:10: fatal error: fftw3.h: No such file or directory
    5 | #include "fftw3.h"
      |          ^~~~~~~~~
compilation terminated.
In file included from /home/sbn/sbndqm_v1_03_00/srcs/sbndqm/sbndqm/dqmAnalysis/TPC/AnalysisSBND.hh:28,
                 from /home/sbn/sbndqm_v1_03_00/srcs/sbndqm/sbndqm/dqmAnalysis/TPC/SBND/OnlineEvdSBND_module.cc:35:
/home/sbn/sbndqm_v1_03_00/srcs/sbndqm/sbndqm/dqmAnalysis/TPC/FFT.hh:5:10: fatal error: fftw3.h: No such file or directory
    5 | #include "fftw3.h"
      |          ^~~~~~~~~
compilation terminated.
```

I just pushed a commit that cleans up the CMakeLists.txt file a little to get it to build for me. (https://github.com/SBNSoftware/sbndqm/pull/39/commits/a2dcc68ceecc4db7e808ffc07e56e6025f4c9f75)

**Review by betan009 — APPROVED** [review]

**Pull request merged**